File: 1046-max-consecutive-ones-iii/1046-max-consecutive-ones-iii.go
func longestOnes(nums []int, k int) int {
    left := 0
    maxLength := 0
    zerosCount := 0

    for right := 0; right < len(nums); right++ {
        if nums[right] == 0 {
            zerosCount++
        }

        for zerosCount > k {
            if nums[left] == 0 {
                zerosCount--
            }
            left++
        }

        maxLength = max(maxLength, right - left + 1)
    }

    return maxLength
}

File: 1046-max-consecutive-ones-iii/1046-max-consecutive-ones-iii.java
public class Solution {
    public int longestOnes(int[] nums, int k) {
        int l = 0;
        int maxLength = 0;
        int zerosCount = 0;

        for (int r = 0; r < nums.length; r++) {
            if (nums[r] == 0) {
                zerosCount++;
            }
            while (zerosCount > k) {
                if (nums[l] == 0) {
                    zerosCount--;
                }
                l++;
            }
            maxLength = Math.max(maxLength, r - l + 1);
        }

        return maxLength;
    }
}
